rust/automerge/src/query/spans.rs

```diff
@@ -0,0 +1,135 @@
+use crate::query::{OpSetMetadata, QueryResult, TreeQuery};
+use crate::types::{ElemId, ListEncoding, MarkName, Op, OpType, ScalarValue};
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct Spans<'a> {
+    pos: usize,
+    seen: usize,
+    encoding: ListEncoding,
+    last_seen: Option<ElemId>,
+    last_insert: Option<ElemId>,
+    seen_at_this_mark: Option<ElemId>,
+    seen_at_last_mark: Option<ElemId>,
+    ops: Vec<&'a Op>,
+    marks: HashMap<MarkName, &'a ScalarValue>,
+    changed: bool,
+    spans: Vec<InternalSpan<'a>>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Span<'a> {
+    pub pos: usize,
+    pub marks: Vec<(smol_str::SmolStr, Cow<'a, ScalarValue>)>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct InternalSpan<'a> {
+    pos: usize,
+    marks: Vec<(MarkName, Cow<'a, ScalarValue>)>,
+}
+
+impl<'a> InternalSpan<'a> {
+    fn into_span(self, m: &OpSetMetadata) -> Span<'a> {
+        Span {
+            pos: self.pos,
+            marks: self
+                .marks
+                .into_iter()
+                .map(|(name, value)| (m.props.get(name.props_index()).into(), value))
+                .collect(),
+        }
+    }
+}
+
+impl<'a> Spans<'a> {
+    pub(crate) fn new(encoding: ListEncoding) -> Self {
+        Spans {
+            pos: 0,
+            seen: 0,
+            encoding,
+            last_seen: None,
+            last_insert: None,
+            seen_at_last_mark: None,
+            seen_at_this_mark: None,
+            changed: false,
+            ops: Vec::new(),
+            marks: HashMap::new(),
+            spans: Vec::new(),
+        }
+    }
+
+    pub(crate) fn into_spans(self, m: &OpSetMetadata) -> Vec<Span<'a>> {
+        self.spans.into_iter().map(|s| s.into_span(m)).collect()
+    }
+
+    pub(crate) fn check_marks(&mut self) {
+        let mut new_marks = HashMap::new();
+        for op in &self.ops {
+            if let OpType::MarkBegin(m) = &op.action {
```

Comment:
:question: Could the ExId be included in `self.marks` as well? For users that use the data from `doc.spans()` to render in an editor, it seems that including the mark ExId would be useful, rather than having to look it up later in `.raw_spans()`.

Reply:
I've made some major changes to the api.  Now the only api calls are 

```typescript
  mark(obj: ObjID, key: string, range: string, value: Value, datatype?: Datatype): void;
  unmark(obj: ObjID, key: string, start: number, end: numbr): void;
  marks(obj: ObjID): Mark[];
```

Unmark now takes a range.  Marks do not have Id's.  The terminology should be consistent.